// @kustomer/apps-server-sdk/lib/constants/roles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.common = exports.custom_kobject = exports.kobject = exports.customer = exports.Permission = exports.User = void 0;
var User;
(function (User) {
    User["CustomerWrite"] = "org.user.customer.write";
    User["CustomerRead"] = "org.user.customer.read";
    User["KObjectRead"] = "org.user.kobject.read";
    User["KObjectWrite"] = "org.user.kobject.write";
})(User = exports.User || (exports.User = {}));
var Permission;
(function (Permission) {
    Permission["CustomerRead"] = "org.permission.customer.read";
    Permission["CustomerCreate"] = "org.permission.customer.create";
    Permission["CustomerUpdate"] = "org.permission.customer.update";
    Permission["KObjectCreate"] = "org.permission.kobject.create";
    Permission["KObjectUpdate"] = "org.permission.kobject.update";
    Permission["CustomKObjectUpdate"] = "org.permission.kobject.kobject_*.update";
    Permission["CustomKObjectCreate"] = "org.permission.kobject.kobject_*.create";
})(Permission = exports.Permission || (exports.Permission = {}));
exports.customer = [
    User.CustomerRead,
    User.CustomerWrite,
    Permission.CustomerRead,
    Permission.CustomerUpdate,
    Permission.CustomerCreate
];
exports.kobject = [
    User.KObjectRead,
    User.KObjectWrite,
    Permission.KObjectCreate,
    Permission.KObjectUpdate
];
exports.custom_kobject = [
    Permission.CustomKObjectCreate,
    Permission.CustomKObjectUpdate
];
exports.common = [...exports.customer, ...exports.kobject, ...exports.custom_kobject];
