// @kustomer/apps-server-sdk/lib/testing/stubs/kapp.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KAppStub = void 0;
const crypto_1 = __importDefault(require("crypto"));
const sinon_1 = __importDefault(require("sinon"));
const kapp_1 = require("../../kapp");
const api = __importStar(require("../../api"));
class KAppStub extends kapp_1.KApp {
    onInstall = jest.fn();
    onEnable = jest.fn();
    onDisable = jest.fn();
    _orgs;
    constructor(options = {}) {
        super({
            app: 'my_test_app',
            version: '1.0.0',
            title: 'My Test App',
            description: 'my test app',
            visibility: 'public',
            url: 'https://my_test_app.com',
            iconUrl: 'https://my_tets_app.com/assets/icon.png',
            clientId: crypto_1.default.randomBytes(8).toString('hex'),
            clientSecret: crypto_1.default.randomBytes(8).toString('hex'),
            roles: [],
            appDetails: {
                appDeveloper: {
                    name: 'Kustomer',
                    supportEmail: 'support@kustomer.com',
                    website: 'https://www.kustomer.com'
                },
                externalPlatform: {
                    name: 'Kustomer',
                    website: 'https://www.kustomer.com'
                }
            },
            ...options
        });
        const orgs = new api.Orgs('token', 'org-id', this.options);
        this._orgs = {
            ...sinon_1.default.stub(orgs),
            customers: sinon_1.default.stub(orgs.customers),
            kobjects: sinon_1.default.stub(orgs.kobjects),
            settings: sinon_1.default.stub(orgs.settings),
            messages: sinon_1.default.stub(orgs.messages),
            log: sinon_1.default.stub(orgs.log)
        };
    }
    start = jest.fn(() => Promise.resolve());
    stop = jest.fn(() => Promise.resolve());
    org(_identifier) {
        return this._orgs;
    }
}
exports.KAppStub = KAppStub;
