// @kustomer/apps-server-sdk/lib/api/base.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIBase = void 0;
const axios_1 = __importDefault(require("axios"));
const constants_1 = require("../constants");
const logger_1 = require("../logger");
// an in memory org name/id => token mapping
const ORG_TOKENS = new Map();
class APIBase {
    token;
    name;
    orgNameOrId;
    options;
    axios;
    _log;
    get baseUrls() {
        return constants_1.ENVS[this.options.env || 'prod'];
    }
    /**
     * @param token the apps token
     * @param name the name of the api
     * @param orgNameOrId the name or id of the org in reference (unique)
     * @param options the apps options
     */
    constructor(token, name, orgNameOrId, options) {
        this.token = token;
        this.name = name;
        this.orgNameOrId = orgNameOrId;
        this.options = options;
        this._log =
            this.options.logger ||
                new logger_1.Logger(`kapp:${this.options.app}:${this.name}`);
        this.axios = axios_1.default.create();
        this.axios.interceptors.request.use(this._onRequest.bind(this));
    }
    /**
     * get an app auth token to make requests in an org
     * @param forceRefresh ignore cached values
     */
    async getToken(forceRefresh) {
        // when token override is given always use it
        if (this.options.token) {
            return this.options.token;
        }
        if (!forceRefresh && ORG_TOKENS.has(this.orgNameOrId)) {
            return ORG_TOKENS.get(this.orgNameOrId);
        }
        try {
            const res = await axios_1.default.get(`${this.baseUrls.gateway}/v1/orgs/${this.orgNameOrId}/token`, {
                headers: {
                    'x-kustomer-app-client-id': this.options.clientId,
                    'x-kustomer-app-token': this.token
                }
            });
            ORG_TOKENS.set(this.orgNameOrId, res.data.data.attributes.token);
            return res.data.data.attributes.token;
        }
        catch (err) {
            this._log.error(err);
        }
    }
    /**
     * @private
     */
    async _onRequest(config) {
        const token = await this.getToken();
        if (!token) {
            throw new Error(`failed to get token for ${this.orgNameOrId}`);
        }
        return {
            ...config,
            headers: {
                ...config.headers,
                Authorization: `Bearer ${token}`
            }
        };
    }
}
exports.APIBase = APIBase;
