// @kustomer/apps-server-sdk/lib/api/customer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Customers = void 0;
const utils_1 = require("../utils");
const base_1 = require("./base");
class Customers extends base_1.APIBase {
    constructor(token, orgNameOrId, options) {
        super(token, 'customers', orgNameOrId, options);
    }
    /**
     * get a customer by id
     * @param id customer id (unique)
     */
    async getById(id) {
        try {
            const res = await this.axios.get(`${this.baseUrls.api}/v1/customers/id=${id}`);
            return {
                id: res.data.data.id,
                ...res.data.data.attributes
            };
        }
        catch (err) {
            if (err.response?.status !== 404) {
                throw err;
            }
        }
    }
    /**
     * get a customer by external id
     * @param externalId customer external id
     */
    async getByExternalId(externalId) {
        try {
            const res = await this.axios.get(`${this.baseUrls.api}/v1/customers/externalId=${externalId}`);
            return {
                id: res.data.data.id,
                ...res.data.data.attributes
            };
        }
        catch (err) {
            if (err.response?.status !== 404) {
                throw err;
            }
        }
    }
    /**
     * get a customer by email
     * @param email customer email (unique)
     */
    async getByEmail(email) {
        try {
            const res = await this.axios.get(`${this.baseUrls.api}/v1/customers/email=${email}`);
            return {
                id: res.data.data.id,
                ...res.data.data.attributes
            };
        }
        catch (err) {
            if (err.response?.status !== 404) {
                throw err;
            }
        }
    }
    /**
     * get a customer by phone
     * @param phone customer phone (unique)
     * @returns
     */
    async getByPhone(phone) {
        try {
            const res = await this.axios.get(`${this.baseUrls.api}/v1/customers/phone=${phone}`);
            return {
                id: res.data.data.id,
                ...res.data.data.attributes
            };
        }
        catch (err) {
            if (err.response?.status !== 404) {
                throw err;
            }
        }
    }
    /**
     * create a new customer
     */
    async create(args) {
        const res = await this.axios.post(`${this.baseUrls.api}/v1/customers`, args);
        return {
            id: res.data.data.id,
            ...res.data.data.attributes
        };
    }
    /**
     * update an existing customer
     * @param id customer id (unique)
     */
    async update(id, args) {
        const res = await this.axios.patch(`${this.baseUrls.api}/v1/customers/${id}`, args);
        return {
            id: res.data.data.id,
            ...res.data.data.attributes
        };
    }
    /**
     * create a kobject for a customer
     * @param id customer id (unique)
     * @param klass klass name
     */
    async createKObject(id, klass, object) {
        const res = await this.axios.post(`${this.baseUrls.api}/v1/customers/${id}/klasses/${(0, utils_1.namespace)(klass, this.options)}`, object);
        return {
            id: res.data.data.id,
            ...res.data.data.attributes
        };
    }
}
exports.Customers = Customers;
