// @kustomer/apps-server-sdk/lib/api/log.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Log = void 0;
const base_1 = require("./base");
class Log extends base_1.APIBase {
    constructor(token, orgNameOrId, options) {
        super(token, 'log', orgNameOrId, options);
    }
    /**
     * @param message
     * @param body
     */
    async info(message, options) {
        return this.log('info', message, options);
    }
    /**
     * @param message
     * @param body
     */
    async warn(message, options) {
        return this.log('warn', message, options);
    }
    /**
     * @param message
     * @param body
     */
    async error(message, options) {
        return this.log('error', message, options);
    }
    /**
     * @param level
     * @param message
     * @param body
     */
    async log(level, message, options) {
        this._log[level](message);
        await this.axios.post(`${this.baseUrls.api}/v1/logs`, {
            level,
            code: `${options?.code || 111}`,
            app: this.options.app,
            name: `${this.options.app}:${level}`,
            message,
            body: options?.body || message
        });
    }
}
exports.Log = Log;
