// @kustomer/apps-server-sdk/lib/api/org.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Orgs = void 0;
const base_1 = require("./base");
const customer_1 = require("./customer");
const kobject_1 = require("./kobject");
const log_1 = require("./log");
const message_1 = require("./message");
const settings_1 = require("./settings");
class Orgs extends base_1.APIBase {
    customers;
    kobjects;
    settings;
    messages;
    log;
    constructor(token, orgNameOrId, options) {
        super(token, 'orgs', orgNameOrId, options);
        this.customers = new customer_1.Customers(token, orgNameOrId, options);
        this.kobjects = new kobject_1.KObjects(token, orgNameOrId, options);
        this.settings = new settings_1.Settings(token, orgNameOrId, options);
        this.messages = new message_1.Messages(token, orgNameOrId, options);
        this.log = new log_1.Log(token, orgNameOrId, options);
    }
    /**
     * get an org by id
     * @param id org id (unique)
     */
    async getById(id) {
        try {
            const res = await this.axios.get(`${this.baseUrls.api}/v1/orgs/${id}`);
            return {
                id: res.data.data.id,
                ...res.data.data.attributes
            };
        }
        catch (err) {
            if (err.response?.status !== 404) {
                throw err;
            }
        }
    }
}
exports.Orgs = Orgs;
